File: front/nft_game_node_prj/wallets/phantom.ts
import * as solanaWeb3 from "@solana/web3.js";
import { SignedMessage } from "../interfaces/signedMessage";
import Wallet from "./wallet";

declare const window: Window &
  typeof globalThis & {
    solana: any;
  };

class PhantomWallet extends Wallet {
  constructor() {
    super();
  }

  async connect(): Promise<solanaWeb3.PublicKey> {
    const resp = await window.solana.connect();
    console.log(resp);
    return new solanaWeb3.PublicKey(resp.publicKey.toString());
  }

  disconnect(): void {
    window.solana.disconnect();
  }

  isInstalled(): Boolean {
    return window.solana && window.solana.isPhantom;
  }

  async signTransaction(
    tx: solanaWeb3.Transaction
  ): Promise<solanaWeb3.Transaction> {
    return await window.solana.signTransaction(tx);
  }

  async signMessage(msg: string): Promise<SignedMessage> {
    const encodedMsg = new TextEncoder().encode(msg);
    return await window.solana.signMessage(encodedMsg, "utf8");
  }
}

export default new PhantomWallet();

File: front/lib/js/phantom.dart
import 'dart:typed_data';

import 'package:js/js.dart';
import 'package:nft_game/js/solana_web3/transaction.dart';
import 'package:nft_game/js/solana_web3/public_key.dart';

@JS('wallet.phantom.connect')
external Future<PublicKey> connect();

@JS('wallet.phantom.disconnect')
external Future<PublicKey> disconnect();

@JS('wallet.phantom.isInstalled')
external bool? isInstalled();

@JS('wallet.phantom.signTransaction')
external Future<String> signTransaction(Transaction tx);

@anonymous
@JS('wallet.phantom.SignedMessage')
class SignedMessage {
  external factory SignedMessage({Uint8List signature, PublicKey publicKey});
  external Uint8List get signature;
  external PublicKey get publicKey;
}

@JS('wallet.phantom.signMessage')
external Future<String> signMessage(String msg);
